driver-core/src/test/functional/com/mongodb/client/model/expressions/AbstractExpressionsFunctionalTest.java

```diff
@@ -0,0 +1,124 @@
+
+    @BeforeEach
+    public void setUp() {
+        getCollectionHelper().drop();
+    }
+
+    @AfterEach
+    public void tearDown() {
+        getCollectionHelper().drop();
+    }
+
+    protected void assertExpression(final Object expected, final Expression expression) {
+        assertExpression(expected, expression, null);
+    }
+
+    protected void assertExpression(@Nullable final Object expected, final Expression expression, @Nullable final String expectedMql) {
+        assertEval(expected, expression);
+
+        if (expectedMql == null) {
+            return;
+        }
+
+        BsonValue expressionValue = ((MqlExpression<?>) expression).toBsonValue(fromProviders(new BsonValueCodecProvider()));
+        BsonValue bsonValue = new BsonDocumentFragmentCodec().readValue(
+                new JsonReader(expectedMql),
+                DecoderContext.builder().build());
+        assertEquals(bsonValue, expressionValue, expressionValue.toString().replace("\"", "'"));
+    }
+
+    private void assertEval(@Nullable final Object expected, final Expression toEvaluate) {
+        BsonValue evaluated = evaluate(toEvaluate);
+        BsonValue expected1 = toBsonValue(expected);
+        assertEquals(expected1, evaluated);
+    }
+
+    protected BsonValue toBsonValue(@Nullable final Object value) {
+        if (value instanceof BsonValue) {
+            return (BsonValue) value;
+        }
+        return new Document("val", value).toBsonDocument().get("val");
+    }
+
+    protected BsonValue evaluate(final Expression toEvaluate) {
+        Bson addFieldsStage = addFields(new Field<>("val", toEvaluate));
+        List<Bson> stages = new ArrayList<>();
+        stages.add(addFieldsStage);
+        List<BsonDocument> results;
+        if (getCollectionHelper().count() == 0) {
+            BsonDocument document = new BsonDocument("val", new BsonString("#invalid string#"));
+            if (serverVersionAtLeast(5, 1)) {
+                Bson documentsStage = new BsonDocument("$documents", new BsonArray(Arrays.asList(document)));
+                stages.add(0, documentsStage);
+                results = getCollectionHelper().aggregateDb(stages);
+            } else {
+                getCollectionHelper().insertDocuments(document);
+                results = getCollectionHelper().aggregate(stages);
+                getCollectionHelper().drop();
+            }
+        } else {
+            results = getCollectionHelper().aggregate(stages);
+        }
+        return results.get(0).get("val");
+    }
+
+    private static class BsonDocumentFragmentCodec extends BsonDocumentCodec {
+        public BsonValue readValue(final BsonReader reader, final DecoderContext decoderContext) {
+            reader.readBsonType();
+            return super.readValue(reader, decoderContext);
+        }
+    }
+
+
+    static <R extends Expression> R ofRem() {
+        // $$REMOVE is intentionally not exposed to users
```

Comment:
Is it possible to eliminate `missing` from the type system? That is, it would not even be mentioned in docs, but only used internally by some of our operations. It is an interesting mechanism, but other languages seem to do fine without it.

It arises only during `get`-like operations, and is usually (but not always) converted to null by ensuing expressions.

Reply:
`has` was used for this purpose, both on map and doc.